Add suggestion for relaxing static lifetime bounds on dyn trait impls in NLL

b-naber opened the pull request:
This PR introduces suggestions for relaxing static lifetime bounds on impls of dyn trait items for NLL similar to what is already available in lexical region diagnostics.

Fixes https://github.com/rust-lang/rust/issues/95701

r? @estebank

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6545414930?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/50e9b6dbcc4b84d318a10106f77e142c834a0755/checks/6545414930/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0308]: mismatched types
   --> /checkout/compiler/rustc_index/src/lib.rs:21:32
    |
19  | / macro_rules! static_assert_size {
20  | |     ($ty:ty, $size:expr) => {
21  | |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
    | |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 24 elements, found one with 16 elements
23  | | }
    | |_- in this expansion of `rustc_data_structures::static_assert_size!`
    |
   ::: compiler/rustc_middle/src/mir/query.rs:345:1
   ::: compiler/rustc_middle/src/mir/query.rs:345:1
    |
345 |   rustc_data_structures::static_assert_size!(ConstraintCategory<'_>, 24);

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6545809255?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/b70c418535687c67a2987256b60346cb4eecdc48/checks/6545809255/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0308]: mismatched types
   --> /checkout/compiler/rustc_index/src/lib.rs:21:32
    |
19  | / macro_rules! static_assert_size {
20  | |     ($ty:ty, $size:expr) => {
21  | |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
    | |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 16 elements, found one with 24 elements
23  | | }
    | |_- in this expansion of `rustc_data_structures::static_assert_size!`
    |
   ::: compiler/rustc_middle/src/mir/query.rs:345:1
   ::: compiler/rustc_middle/src/mir/query.rs:345:1
    |
345 |   rustc_data_structures::static_assert_size!(ConstraintCategory<'_>, 16);

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

b-naber commented:
Weird that this compiled locally for me with the start_assert_size of 24.

Probably needs a perf-run.

jackh726 commented:
I would have expected the NLL output of `src/test/ui/suggestions/impl-on-dyn-trait-with-implicit-static-bound.rs` to change. Did you bless with compare-mode=nll?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6551093484?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/1fda1fb72ce585df3cb7758af8cfd748ebfb6dd5/checks/6551093484/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] src/test/ui/associated-types/cache/project-fn-ret-invariant-nll.rs#oneuse stdout ----
diff of stderr:

15    = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
17 error: lifetime may not live long enough
-   --> $DIR/project-fn-ret-invariant-nll.rs:47:13
+   --> $DIR/project-fn-ret-invariant-nll.rs:46:13
19    |
19    |
20 LL | fn baz<'a, 'b>(x: Type<'a>, y: Type<'b>) -> (Type<'a>, Type<'b>) {
21    |        --  -- lifetime `'b` defined here
22    |        |
22    |        |
23    |        lifetime `'a` defined here
- ...
- LL |     let b = bar(f, y);
+ LL |     let f = foo; // <-- No consistent type can be inferred for `f` here.
+ LL |     let a = bar(f, x);
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
26    |             ^^^^^^^^^ argument requires that `'b` must outlive `'a`
27    |
28    = help: consider adding the following bound: `'b: 'a`

-    = note: requirement occurs because of the type `Type<'_>`, which makes the generic argument `'_` invariant
-    = note: the struct `Type<'a>` is invariant over the parameter `'a`
+    = note: requirement occurs because of a function pointer to `foo`
+    = note: the function `foo` is invariant over the parameter `'a`
31    = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
32 
33 help: `'a` and `'b` must be the same: replace one with the other

The actual stderr differed from the expected stderr.
The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/cache/project-fn-ret-invariant-nll.oneuse/project-fn-ret-invariant-nll.oneuse.stderr
To only update this specific test, also pass `--test-args associated-types/cache/project-fn-ret-invariant-nll.rs`


error in revision `oneuse`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/associated-types/cache/project-fn-ret-invariant-nll.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "oneuse" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/cache/project-fn-ret-invariant-nll.oneuse" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "borrowck=mir" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/associated-types/cache/project-fn-ret-invariant-nll.oneuse/auxiliary"
stdout: none
--- stderr -------------------------------
error: lifetime may not live long enough
   |
   |
LL | fn baz<'a, 'b>(x: Type<'a>, y: Type<'b>) -> (Type<'a>, Type<'b>) {
   |        --  -- lifetime `'b` defined here
   |        |
   |        lifetime `'a` defined here
LL |     let f = foo; // <-- No consistent type can be inferred for `f` here.
LL |     let a = bar(f, x); //[oneuse]~ ERROR lifetime may not live long enough
   |             ^^^^^^^^^ argument requires that `'a` must outlive `'b`
   |
   = help: consider adding the following bound: `'a: 'b`
   = note: requirement occurs because of the type `Type<'_>`, which makes the generic argument `'_` invariant
   = note: the struct `Type<'a>` is invariant over the parameter `'a`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
error: lifetime may not live long enough
  --> /checkout/src/test/ui/associated-types/cache/project-fn-ret-invariant-nll.rs:46:13
   |
   |
LL | fn baz<'a, 'b>(x: Type<'a>, y: Type<'b>) -> (Type<'a>, Type<'b>) {
   |        --  -- lifetime `'b` defined here
   |        |
   |        lifetime `'a` defined here
LL |     let f = foo; // <-- No consistent type can be inferred for `f` here.
LL |     let a = bar(f, x); //[oneuse]~ ERROR lifetime may not live long enough
   |             ^^^^^^^^^ argument requires that `'b` must outlive `'a`
   |
   = help: consider adding the following bound: `'b: 'a`
   = note: requirement occurs because of a function pointer to `foo`
   = note: the function `foo` is invariant over the parameter `'a`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

help: `'a` and `'b` must be the same: replace one with the other
error: aborting due to 2 previous errors
------------------------------------------



---- [ui] src/test/ui/async-await/issues/issue-72312.rs#nll stdout ----
diff of stderr:

1 error[E0521]: borrowed data escapes outside of associated function
-   --> $DIR/issue-72312.rs:20:24
3    |
4 LL |       pub async fn start(&self) {
5    |                          -----


7    |                          `self` is a reference that is only valid in the associated function body
8    |                          let's call the lifetime of this reference `'1`
9 ...
- LL |           require_static(async move {
-    |  ________________________^
+ LL | /         require_static(async move {
12 LL | |
13 LL | |
14 LL | |
15 LL | |
16 LL | |             &self;
17 LL | |         });
-    | |         ^
-    | |         ^
-    | |         |
-    | |_________`self` escapes the associated function body here
-    |           argument requires that `'1` must outlive `'static`
+    | |          |
+    | |          |
+    | |__________`self` escapes the associated function body here
+    |            argument requires that `'1` must outlive `'static`
23 error: aborting due to previous error
24 



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issues/issue-72312.nll/issue-72312.nll.stderr
To only update this specific test, also pass `--test-args async-await/issues/issue-72312.rs`


error in revision `nll`: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/async-await/issues/issue-72312.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--cfg" "nll" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issues/issue-72312.nll" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zborrowck=mir" "--edition=2018" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/async-await/issues/issue-72312.nll/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0521]: borrowed data escapes outside of associated function
   |
LL |       pub async fn start(&self) {
   |                          -----
   |                          |
   |                          |
   |                          `self` is a reference that is only valid in the associated function body
   |                          let's call the lifetime of this reference `'1`
...
LL | /         require_static(async move {
LL | |             //[base]~^ NOTE ...and is required to live as long as `'static` here
LL | |             //[nll]~^^ ERROR borrowed data escapes
LL | |             //[nll]~| NOTE `self` escapes
LL | |             //[nll]~| NOTE argument requires
LL | |             &self; //[base]~ NOTE ...is used here...
LL | |         });
   | |          |
   | |          |
   | |__________`self` escapes the associated function body here
   |            argument requires that `'1` must outlive `'static`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0521`.
------------------------------------------
------------------------------------------


---- [ui] src/test/ui/suggestions/impl-on-dyn-trait-with-implicit-static-bound-nll.rs stdout ----
diff of stderr:

12    |         argument requires that `'a` must outlive `'static`
13    |
14 note: the used `impl` has a `'static` requirement
-   --> $DIR/impl-on-dyn-trait-with-implicit-static-bound.rs:14:32
16    |
16    |
17 LL |     impl<T> MyTrait<T> for dyn ObjectTrait<T> {
18    |                                ^^^^^^^^^^^^^^ this has an implicit `'static` lifetime requirement

37    |         argument requires that `'a` must outlive `'static`
38    |
39 note: the used `impl` has a `'static` requirement
-   --> $DIR/impl-on-dyn-trait-with-implicit-static-bound.rs:64:14
41    |
41    |
42 LL |     impl dyn ObjectTrait {
43    |              ^^^^^^^^^^^ this has an implicit `'static` lifetime requirement

62    |         argument requires that `'a` must outlive `'static`
63    |
64 note: the used `impl` has a `'static` requirement
-   --> $DIR/impl-on-dyn-trait-with-implicit-static-bound.rs:85:26
66    |
66    |
67 LL |         fn use_self(&self) -> &() { panic!() }
68    |            -------- calling this method introduces the `impl`'s 'static` requirement

88    |         argument requires that `'a` must outlive `'static`
89    |
90 note: the used `impl` has a `'static` requirement
-   --> $DIR/impl-on-dyn-trait-with-implicit-static-bound.rs:104:26
92    |
92    |
93 LL |         fn use_self(&self) -> &() { panic!() }
94    |            -------- calling this method introduces the `impl`'s 'static` requirement

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/impl-on-dyn-trait-with-implicit-static-bound-nll/impl-on-dyn-trait-with-implicit-static-bound-nll.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args suggestions/impl-on-dyn-trait-with-implicit-static-bound-nll.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/suggestions/impl-on-dyn-trait-with-implicit-static-bound-nll.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/impl-on-dyn-trait-with-implicit-static-bound-nll" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Zborrowck=mir" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/suggestions/impl-on-dyn-trait-with-implicit-static-bound-nll/auxiliary"
stdout: none
--- stderr -------------------------------
error[E0521]: borrowed data escapes outside of function
   |
   |
LL |     fn use_it<'a, T>(val: &'a dyn ObjectTrait<T>) -> impl OtherTrait<'a> + 'a {
   |               --     --- `val` is a reference that is only valid in the function body
   |               |
   |               lifetime `'a` defined here
LL |         val.use_self::<T>() //~ ERROR borrowed data escapes
   |         |
   |         |
   |         `val` escapes the function body here
   |         argument requires that `'a` must outlive `'static`
   |
note: the used `impl` has a `'static` requirement
   |
   |
LL |     impl<T> MyTrait<T> for dyn ObjectTrait<T> {
   |                                ^^^^^^^^^^^^^^ this has an implicit `'static` lifetime requirement
LL |         fn use_self<K>(&self) -> &() { panic!() }
   |            -------- calling this method introduces the `impl`'s 'static` requirement
help: consider relaxing the implicit `'static` requirement
   |
LL |     impl<T> MyTrait<T> for dyn ObjectTrait<T> + '_ {


error[E0521]: borrowed data escapes outside of function
   |
   |
LL |     fn use_it<'a>(val: &'a dyn ObjectTrait) -> impl OtherTrait<'a> + 'a {
   |               --  --- `val` is a reference that is only valid in the function body
   |               |
   |               lifetime `'a` defined here
LL |         val.use_self()
   |         |
   |         |
   |         `val` escapes the function body here
   |         argument requires that `'a` must outlive `'static`
   |
note: the used `impl` has a `'static` requirement
   |
   |
LL |     impl dyn ObjectTrait {
   |              ^^^^^^^^^^^ this has an implicit `'static` lifetime requirement
LL |         fn use_self(&self) -> &() { panic!() }
   |            -------- calling this method introduces the `impl`'s 'static` requirement
help: consider relaxing the implicit `'static` requirement
   |
LL |     impl dyn ObjectTrait + '_ {


error[E0521]: borrowed data escapes outside of function
   |
   |
LL |     fn use_it<'a>(val: &'a dyn ObjectTrait) -> impl OtherTrait<'a> {
   |               --  --- `val` is a reference that is only valid in the function body
   |               |
   |               lifetime `'a` defined here
LL |         val.use_self() //~ ERROR borrowed data escapes
   |         |
   |         |
   |         `val` escapes the function body here
   |         argument requires that `'a` must outlive `'static`
   |
note: the used `impl` has a `'static` requirement
   |
   |
LL |         fn use_self(&self) -> &() { panic!() }
   |            -------- calling this method introduces the `impl`'s 'static` requirement
...
LL |     impl MyTrait for dyn ObjectTrait {}
   |                          ^^^^^^^^^^^ this has an implicit `'static` lifetime requirement
help: consider relaxing the implicit `'static` requirement
   |
LL |     impl MyTrait for dyn ObjectTrait + '_ {}


error[E0521]: borrowed data escapes outside of function
   |
   |
LL |     fn use_it<'a>(val: &'a dyn ObjectTrait) -> impl OtherTrait<'a> + 'a {
   |               --  --- `val` is a reference that is only valid in the function body
   |               |
   |               lifetime `'a` defined here
LL |         MyTrait::use_self(val) //~ ERROR borrowed data escapes
   |         |
   |         |
   |         `val` escapes the function body here
   |         argument requires that `'a` must outlive `'static`
   |
note: the used `impl` has a `'static` requirement
   |
   |
LL |         fn use_self(&self) -> &() { panic!() }
   |            -------- calling this method introduces the `impl`'s 'static` requirement
...
LL |     impl MyTrait for dyn ObjectTrait {}
   |                          ^^^^^^^^^^^ this has an implicit `'static` lifetime requirement
help: consider relaxing the implicit `'static` requirement
   |
LL |     impl MyTrait for dyn ObjectTrait + '_ {}

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0521`.
```

</details>

jackh726 commented on compiler/rustc_middle/src/ty/impls_ty.rs:
```diff
@@ -135,8 +135,8 @@ impl<'a> HashStable<StableHashingContext<'a>> for ty::RegionKind {
             }
-            ty::ReVar(..) => {
-                bug!("StableHasher: unexpected region {:?}", *self)
+            ty::ReVar(reg_vid) => {
+                reg_vid.hash_stable(hcx, hasher);
```
Why did this change?

b-naber replied:
Was hitting an ICE because `ConstraintCategory` is `HashStable` and we encounter `ReVar` in the `substs` that we added.

b-naber replied:
Wouldn't we still ICE if we added that here (due to delay_span_bug being called, but no error having been found)?

estebank replied:
Yes, but only if no other errors are emitted, which would only happen if a diagnostic isn't built. I'm assuming that `ReVar` is only being encountered due to diagnostic changes and not changes that would affect the final binary, right?

jackh726 commented on compiler/rustc_middle/src/mir/query.rs:
```diff
@@ -338,20 +338,17 @@ pub struct ClosureOutlivesRequirement<'tcx> {
+    pub category: ConstraintCategory<'tcx>,
 }
 
-// Make sure this enum doesn't unintentionally grow
-rustc_data_structures::static_assert_size!(ConstraintCategory, 12);
-
```
Why remove the assert? (Are there problems when there is a lifetime?)

b-naber replied:
Because `SubstsRef` doesn't have a static size I think, at least I was hitting that assertion.

b-naber replied:
Eh that's nonsense, of course it must have a static size. I really don't know why I was hitting those assertions, but the first two CI runs failed because of that, didn't have that problem locally.

jackh726 commented on compiler/rustc_infer/src/infer/error_reporting/nice_region_error/util.rs:
```diff
@@ -51,9 +52,19 @@ pub fn find_param_with_region<'tcx>(
     let body_id = hir.maybe_body_owned_by(hir_id)?;
     let body = hir.body(body_id);
+
+    // Don't perform this on closures
```
*why*?

b-naber replied:
Because we can't call  `fn_sig` on closures.

estebank replied:
In the resulting ICE if you call `fn_sig` on closures, it tells you what you should call instead: `substs.as_closure().sig()`

b-naber replied:
I don't think it actually makes sense to look for closures here given the function's intention. This was previously never called on closures, it's just that the way we can call this we can still end up with a closure in this call and we need the HIR check to properly reject this if I remember correctly.

estebank replied:
I suspect that this might have been what caused the output changes that used to point at `async` and closures (but don't know for sure and the output changes aren't too dramatic).

jackh726 commented on compiler/rustc_borrowck/src/diagnostics/region_errors.rs:
```diff
@@ -699,6 +775,99 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
+        let mut visitor = TraitObjectVisitor(FxHashSet::default());
+        visitor.visit_ty(param.param_ty);
+
+        if let Some((ident, self_ty)) =
+            self.get_impl_ident_and_self_ty_from_trait(instance.def_id(), &visitor.0)
+        {
```
Could use `let else`

jackh726 reviewed (changes requested):
Still need to look at this a bit more, but was just thinking if there was any way to unify these suggestions with the existing ones?

bors commented:
:umbrella: The latest upstream changes (presumably #96098) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank commented on compiler/rustc_borrowck/src/diagnostics/region_errors.rs:
```diff
@@ -256,6 +263,70 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         outlives_suggestion.add_suggestion(self);
     }
 
+    fn get_impl_ident_and_self_ty_from_trait(
```
It feels to me like we might already have code doing this somewhere 🤔

b-naber replied:
We do already have that functionality in lexical region diagnostics (`NiceRegionError`), but can't directly use that struct in NLL (and at some point we will get rid of it anyway afaict).

estebank commented on compiler/rustc_borrowck/src/diagnostics/region_errors.rs:
```diff
@@ -699,6 +775,99 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
+            let mut hir_v = HirTraitObjectVisitor(&mut traits, *found_did);
+            hir_v.visit_ty(&self_ty);
+            debug!("trait spans found: {:?}", traits);
+            for span in &traits {
+                let mut multi_span: MultiSpan = vec![*span].into();
+                multi_span.push_span_label(
+                    *span,
+                    "this has an implicit `'static` lifetime requirement".to_string(),
+                );
+                multi_span.push_span_label(
+                    ident.span,
+                    "calling this method introduces the `impl`'s 'static` requirement".to_string(),
+                );
+                err.span_note(multi_span, "the used `impl` has a `'static` requirement");
+                err.span_suggestion_verbose(
+                    span.shrink_to_hi(),
+                    "consider relaxing the implicit `'static` requirement",
+                    " + '_".to_string(),
+                    Applicability::MaybeIncorrect,
+                );
+                suggested = true;
+            }
```
I also think that we might already have logic for this, but if it is not accessible for some reason, I'm not above a little duplication for a clear diagnostic improvement.

estebank commented on src/test/ui/nll/closure-requirements/propagate-approximated-shorter-to-static-comparing-against-free.stderr:
```diff
@@ -72,14 +72,15 @@ LL | | }
-   |                argument requires that `a` is borrowed for `'static`
+LL |       let cell = Cell::new(&a);
+   |                            ^^ borrowed value does not live long enough
 ...
-LL | }
-   | - `a` dropped here while still borrowed
+LL | /     foo(cell, |cell_a, cell_x| {
+LL | |         cell_x.set(cell_a.get()); // forces 'a: 'x, implies 'a = 'static -> borrow error
+LL | |     })
+   | |______- argument requires that `a` is borrowed for `'static`
+LL |   }
+   |   - `a` dropped here while still borrowed
```
I'm somewhat worried about the output changes here and in the `async` block in `/issue-72312.nll.stderr`. It also seems like we should also special case their output if we could 🤔

estebank replied:
*This* case actually seems like an improvement, we "just" need to figure out how to point at the inside of the closure :-/

b-naber replied:
It's kind of weird that this does not point to the start of the closure anymore. Personally I don't really think this makes the diagnostic any less helpful (since the point where the drop happen implies that anyway), but I can try to investigate if you find this important.

estebank replied:
Would you mind to do so as a follow up? In this case in particular it is an improvement, but I'm not so sure about the other ones.

b-naber commented on compiler/rustc_borrowck/src/type_check/mod.rs:
```diff
@@ -1582,11 +1603,25 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
+        } else {
+            None
+        };
+        debug!(?call_arg);
```
Yes I believe so.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6600561104?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/5422fea3fbd363eb62d7513c10d46b85cd84c8f5/checks/6600561104/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_ast_lowering v0.0.0 (/checkout/compiler/rustc_ast_lowering)
   Compiling rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
   Compiling rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error: unused variable: `reg_vid`
   --> compiler/rustc_middle/src/ty/impls_ty.rs:138:23
    |
138 |             ty::ReVar(reg_vid) => {
    |                       ^^^^^^^ help: if this is intentional, prefix it with an underscore: `_reg_vid`
    |
    = note: `-D unused-variables` implied by `-D warnings`
error: could not compile `rustc_middle` due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:01:45
```

</details>

b-naber commented:
@estebank Addressed your review. Use the `Ty` of the function directly now, instead of its `DefId` + substs, which does still increase the size of `ConstraintCategory` by 4 bytes unfortunately.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6600697247?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/d3be76447f2845d48007d9bb712ccb20bd61e533/checks/6600697247/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0308]: mismatched types
   --> /checkout/compiler/rustc_index/src/lib.rs:21:32
    |
19  | / macro_rules! static_assert_size {
20  | |     ($ty:ty, $size:expr) => {
21  | |         const _: [(); $size] = [(); ::std::mem::size_of::<$ty>()];
    | |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 16 elements, found one with 12 elements
23  | | }
    | |_- in this expansion of `rustc_data_structures::static_assert_size!`
    |
   ::: compiler/rustc_middle/src/mir/query.rs:345:1
   ::: compiler/rustc_middle/src/mir/query.rs:345:1
    |
345 |   rustc_data_structures::static_assert_size!(ConstraintCategory<'_>, 16);

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle` due to previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

estebank commented on compiler/rustc_borrowck/src/type_check/mod.rs:
```diff
@@ -1585,11 +1602,20 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             span_mirbug!(self, term, "call to {:?} with wrong # of args", sig);
         }
+
+        let func_ty = if let TerminatorKind::Call { func, .. } = &term.kind {
+            Some(func.ty(body, self.infcx.tcx))
+        } else {
+            None
+        };
```
This is slightly shorter:

```suggestion
        let func_ty = match &term.kind {
            TerminatorKind::Call { func, .. } => Some(func.ty(body, self.infcx.tcx)),
            _ => None,
        };
```

estebank commented on compiler/rustc_middle/src/mir/query.rs:
```diff
@@ -360,7 +364,9 @@ pub enum ConstraintCategory<'tcx> {
-    CallArgument(Option<(DefId, SubstsRef<'tcx>)>),
+
+    /// Contains the function type if available.
+    CallArgument(Option<Ty<'tcx>>),
```
I wonder if we can get the static size back down by changing this to be `CallArgument(Box<Option<Ty<'tcx>>)`, because `Ty<'tcx>` is 40 bytes, and I don't know for sure whether boxing will help or not in that case.

b-naber replied:
We can't afaict, we need `ConstraintCategory` to be `Copy`.

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 3c6c8d5a8dbf4db20450ed5793ef35f29c13466c with merge 66284dab23b66746d03a23cbcfb37d354eaa83bf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"3c6c8d5a8dbf4db20450ed5793ef35f29c13466c","merge_sha":"66284dab23b66746d03a23cbcfb37d354eaa83bf"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6615077537?check_suite_focus=true)
Build commit: 66284dab23b66746d03a23cbcfb37d354eaa83bf (`66284dab23b66746d03a23cbcfb37d354eaa83bf`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6615077537?check_suite_focus=true"},"merge_sha":"66284dab23b66746d03a23cbcfb37d354eaa83bf"} -->

rust-timer commented:
Queued 66284dab23b66746d03a23cbcfb37d354eaa83bf with parent 1ab98933fa75d72e882b86feac1a0be3a5b02cb0, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1ab98933fa75d72e882b86feac1a0be3a5b02cb0&end=66284dab23b66746d03a23cbcfb37d354eaa83bf).

rust-timer commented:
Finished benchmarking commit (66284dab23b66746d03a23cbcfb37d354eaa83bf): [comparison url](https://perf.rust-lang.org/compare.html?start=1ab98933fa75d72e882b86feac1a0be3a5b02cb0&end=66284dab23b66746d03a23cbcfb37d354eaa83bf&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1ab98933fa75d72e882b86feac1a0be3a5b02cb0&end=66284dab23b66746d03a23cbcfb37d354eaa83bf&stat=instructions:u)

- Primary benchmarks: 🎉 relevant improvements found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | -0.3%    | -0.3% | 2         |
| Improvements 🎉 <br /> (secondary) | -2.0%    | -5.8% | 19        |
| All 😿🎉 (primary) | -0.3%    | -0.3% | 2         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1ab98933fa75d72e882b86feac1a0be3a5b02cb0&end=66284dab23b66746d03a23cbcfb37d354eaa83bf&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regression found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.5%     | 2.5% | 1         |
| Regressions 😿 <br /> (secondary) | 2.0%     | 2.3% | 2         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.4%    | -3.4% | 1         |
| All 😿🎉 (primary) | 2.5%     | 2.5% | 1         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1ab98933fa75d72e882b86feac1a0be3a5b02cb0&end=66284dab23b66746d03a23cbcfb37d354eaa83bf&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: no relevant changes found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.6%    | -3.7% | 6         |
| All 😿🎉 (primary) | N/A      | N/A | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

estebank commented:
Ok, I guess disregard the `Box` conversation then 👀

estebank commented:
@bors r+

bors commented:
:pushpin: Commit 3c6c8d5a8dbf4db20450ed5793ef35f29c13466c has been approved by `estebank`

<!-- @bors r=estebank 3c6c8d5a8dbf4db20450ed5793ef35f29c13466c -->
<!-- homu: {"type":"Approved","sha":"3c6c8d5a8dbf4db20450ed5793ef35f29c13466c","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit 3c6c8d5a8dbf4db20450ed5793ef35f29c13466c with merge 2a0c6ac83a9a29409d6db4d5f2d77457261993ce...
<!-- homu: {"type":"BuildStarted","head_sha":"3c6c8d5a8dbf4db20450ed5793ef35f29c13466c","merge_sha":"2a0c6ac83a9a29409d6db4d5f2d77457261993ce"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6618860813?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6618860813?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6618628717?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/2a0c6ac83a9a29409d6db4d5f2d77457261993ce/checks/6618628717/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

---- [ui] src/test/ui/invalid/invalid-llvm-passes.rs stdout ----
diff of stderr:

1 error: failed to run LLVM passes: unknown pass name 'unknown-pass'
+ thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
+   left: `LLVMing`,
+   left: `LLVMing`,
+  right: `Codegenning`', $COMPILER_DIR/rustc_codegen_ssa/src/back/write.rs:1471:21
+    0:     0x7fc5b0a9818d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha6d42905ae8495e1
+    1:     0x7fc5b0af401c - core::fmt::write::he61c0a83568f530b
+    1:     0x7fc5b0af401c - core::fmt::write::he61c0a83568f530b
+    2:     0x7fc5b0a89921 - std::io::Write::write_fmt::hf12b960ff621bcd4
+    3:     0x7fc5b0a9ae75 - std::panicking::default_hook::{{closure}}::h1e32ea7d973577fd
+    4:     0x7fc5b0a9ab96 - std::panicking::default_hook::h0f733119b617cb2f
+    5:     0x7fc5b1872b19 - rustc_driver[956148c2c5381e18]::DEFAULT_HOOK::{closure#0}::{closure#0}
Some tests failed in compiletest suite=ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-musl
+    6:     0x7fc5b0a9b54a - std::panicking::rust_panic_with_hook::h10f2619916c4d7f0
+    7:     0x7fc5b0a9b387 - std::panicking::begin_panic_handler::{{closure}}::hc10db1e63006eff3
+    8:     0x7fc5b0a98644 - std::sys_common::backtrace::__rust_end_short_backtrace::h8514961be8260c05
+    9:     0x7fc5b0a9b0b9 - rust_begin_unwind
+   10:     0x7fc5b0a60413 - core::panicking::panic_fmt::h5cbf0cc0bcd5482c
+   11:     0x7fc5b0af0dc8 - core::panicking::assert_failed_inner::h0cd989a79f6863cc
+   12:     0x7fc5b15e43bb - core[b767cee21b229350]::panicking::assert_failed::<rustc_codegen_ssa[b548771e8a124ff8]::back::write::MainThreadWorkerState, rustc_codegen_ssa[b548771e8a124ff8]::back::write::MainThreadWorkerState>
+   13:     0x7fc5b1ada995 - std[8a7407eec8ab889b]::sys_common::backtrace::__rust_begin_short_backtrace::<<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend as rustc_codegen_ssa[b548771e8a124ff8]::traits::backend::ExtraBackendMethods>::spawn_thread<rustc_codegen_ssa[b548771e8a124ff8]::back::write::start_executing_work<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend>::{closure#4}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>::{closure#0}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>
+   14:     0x7fc5b1b4cfe4 - <<std[8a7407eec8ab889b]::thread::Builder>::spawn_unchecked_<<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend as rustc_codegen_ssa[b548771e8a124ff8]::traits::backend::ExtraBackendMethods>::spawn_thread<rustc_codegen_ssa[b548771e8a124ff8]::back::write::start_executing_work<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend>::{closure#4}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>::{closure#0}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>::{closure#1} as core[b767cee21b229350]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
+   15:     0x7fc5b0aa5463 - std::sys::unix::thread::Thread::new::thread_start::h6cfee43c7d984174
+   16:     0x7fc5b07e4609 - start_thread
+   17:     0x7fc5b0926133 - clone
+   18:                0x0 - <unknown>
+ error: internal compiler error: unexpected panic
+ 
+ note: the compiler unexpectedly panicked. this is a bug.
+ 
+ 
+ note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
+ 
+ note: rustc 1.63.0-nightly (2a0c6ac83 2022-05-27) running on x86_64-unknown-linux-gnu
+ 
+ note: compiler flags: -Z threads=1 -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -C prefer-dynamic -C rpath -C debuginfo=0 -C linker=x86_64-linux-musl-gcc -C passes=unknown-pass -Z new-llvm-pass-manager=yes
+ query stack during panic:
+ end of query stack
3 error: aborting due to previous error
4 
---
To only update this specific test, also pass `--test-args invalid/invalid-llvm-passes.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/invalid/invalid-llvm-passes.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-musl" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-llvm-passes" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-musl/native/rust-test-helpers" "-Clinker=x86_64-linux-musl-gcc" "-Cpasses=unknown-pass" "-Z" "new-llvm-pass-manager=yes" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/invalid/invalid-llvm-passes/auxiliary"
stdout: none
--- stderr -------------------------------
error: failed to run LLVM passes: unknown pass name 'unknown-pass'
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `LLVMing`,
  left: `LLVMing`,
 right: `Codegenning`', /checkout/compiler/rustc_codegen_ssa/src/back/write.rs:1471:21
   0:     0x7fc5b0a9818d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::ha6d42905ae8495e1
   1:     0x7fc5b0af401c - core::fmt::write::he61c0a83568f530b
   1:     0x7fc5b0af401c - core::fmt::write::he61c0a83568f530b
   2:     0x7fc5b0a89921 - std::io::Write::write_fmt::hf12b960ff621bcd4
   3:     0x7fc5b0a9ae75 - std::panicking::default_hook::{{closure}}::h1e32ea7d973577fd
   4:     0x7fc5b0a9ab96 - std::panicking::default_hook::h0f733119b617cb2f
   5:     0x7fc5b1872b19 - rustc_driver[956148c2c5381e18]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:     0x7fc5b0a9b54a - std::panicking::rust_panic_with_hook::h10f2619916c4d7f0
   7:     0x7fc5b0a9b387 - std::panicking::begin_panic_handler::{{closure}}::hc10db1e63006eff3
   8:     0x7fc5b0a98644 - std::sys_common::backtrace::__rust_end_short_backtrace::h8514961be8260c05
   9:     0x7fc5b0a9b0b9 - rust_begin_unwind
  10:     0x7fc5b0a60413 - core::panicking::panic_fmt::h5cbf0cc0bcd5482c
  11:     0x7fc5b0af0dc8 - core::panicking::assert_failed_inner::h0cd989a79f6863cc
  12:     0x7fc5b15e43bb - core[b767cee21b229350]::panicking::assert_failed::<rustc_codegen_ssa[b548771e8a124ff8]::back::write::MainThreadWorkerState, rustc_codegen_ssa[b548771e8a124ff8]::back::write::MainThreadWorkerState>
  13:     0x7fc5b1ada995 - std[8a7407eec8ab889b]::sys_common::backtrace::__rust_begin_short_backtrace::<<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend as rustc_codegen_ssa[b548771e8a124ff8]::traits::backend::ExtraBackendMethods>::spawn_thread<rustc_codegen_ssa[b548771e8a124ff8]::back::write::start_executing_work<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend>::{closure#4}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>::{closure#0}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>
  14:     0x7fc5b1b4cfe4 - <<std[8a7407eec8ab889b]::thread::Builder>::spawn_unchecked_<<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend as rustc_codegen_ssa[b548771e8a124ff8]::traits::backend::ExtraBackendMethods>::spawn_thread<rustc_codegen_ssa[b548771e8a124ff8]::back::write::start_executing_work<rustc_codegen_llvm[6af3a2890704cafe]::LlvmCodegenBackend>::{closure#4}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>::{closure#0}, core[b767cee21b229350]::result::Result<rustc_codegen_ssa[b548771e8a124ff8]::back::write::CompiledModules, ()>>::{closure#1} as core[b767cee21b229350]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  15:     0x7fc5b0aa5463 - std::sys::unix::thread::Thread::new::thread_start::h6cfee43c7d984174
  16:     0x7fc5b07e4609 - start_thread
  17:     0x7fc5b0926133 - clone
  18:                0x0 - <unknown>
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.63.0-nightly (2a0c6ac83 2022-05-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -C prefer-dynamic -C rpath -C debuginfo=0 -C linker=x86_64-linux-musl-gcc -C passes=unknown-pass -Z new-llvm-pass-manager=yes
query stack during panic:
end of query stack
error: aborting due to previous error
------------------------------------------
```

</details>

estebank commented on compiler/rustc_middle/src/ty/impls_ty.rs:
```diff
@@ -135,8 +135,8 @@ impl<'a> HashStable<StableHashingContext<'a>> for ty::RegionKind {
-            ty::ReVar(..) => {
-                bug!("StableHasher: unexpected region {:?}", *self)
+            ty::ReVar(reg) => {
+                reg.hash_stable(hcx, hasher);
```
I wonder if the LLVM failure  is due to this. As a refresher, what are the failures when you keep the `bug!`?

b-naber replied:
Don't think it's related to that. `ReVar` is `HashStable`, we just previously had the `bug!` call in there, because it wasn't used when hashing `Region` afaict. I just got the output of the message in the bug statement.

I'm sort of leaning towards thinking that this failure is spurious. The test is in the ui tests and passed on all other platforms and the code changes in this PR don't touch anything that should cause different behaviour on different platforms.

jackh726 commented:
@bors retry maybe spurious?

bors commented:
:hourglass: Testing commit 3c6c8d5a8dbf4db20450ed5793ef35f29c13466c with merge 6c557271b8989188a4d0dbb93ad2a7fb1fbe5626...
<!-- homu: {"type":"BuildStarted","head_sha":"3c6c8d5a8dbf4db20450ed5793ef35f29c13466c","merge_sha":"6c557271b8989188a4d0dbb93ad2a7fb1fbe5626"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6632988219?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/6632988219?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-apple-alt`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/6632385556?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/6c557271b8989188a4d0dbb93ad2a7fb1fbe5626/checks/6632385556/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

jackh726 commented:
@bors retry spurious network error

bors commented:
:hourglass: Testing commit 3c6c8d5a8dbf4db20450ed5793ef35f29c13466c with merge ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491...
<!-- homu: {"type":"BuildStarted","head_sha":"3c6c8d5a8dbf4db20450ed5793ef35f29c13466c","merge_sha":"ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6635261975?check_suite_focus=true)
Approved by: estebank
Pushing ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"estebank","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6635261975?check_suite_focus=true"},"merge_sha":"ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491"} -->

rust-timer commented:
Finished benchmarking commit (ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491): [comparison url](https://perf.rust-lang.org/compare.html?start=764b8615e9149431d8790e3c07cb663642fe393d&end=ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=764b8615e9149431d8790e3c07cb663642fe393d&end=ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491&stat=instructions:u)

- Primary benchmarks: 🎉 relevant improvement found
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | N/A      | N/A | 0         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | -0.2%    | -0.2% | 1         |
| Improvements 🎉 <br /> (secondary) | -5.3%    | -5.9% | 6         |
| All 😿🎉 (primary) | -0.2%    | -0.2% | 1         |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=764b8615e9149431d8790e3c07cb663642fe393d&end=ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: mixed results
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 3.1%     | 4.0% | 2         |
| Regressions 😿 <br /> (secondary) | 1.9%     | 2.4% | 5         |
| Improvements 🎉 <br /> (primary) | -4.3%    | -4.3% | 1         |
| Improvements 🎉 <br /> (secondary) | -2.2%    | -2.2% | 1         |
| All 😿🎉 (primary) | 0.6%     | -4.3% | 3         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=764b8615e9149431d8790e3c07cb663642fe393d&end=ed76b773b57cf0aa48ec4e2fc6d6a3f7a9079491&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: mixed results
- Secondary benchmarks: 🎉 relevant improvements found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 2.6%     | 2.6% | 1         |
| Regressions 😿 <br /> (secondary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (primary) | -2.3%    | -2.3% | 1         |
| Improvements 🎉 <br /> (secondary) | -3.5%    | -4.0% | 6         |
| All 😿🎉 (primary) | 0.2%     | 2.6% | 2         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
